proposals/0000-swift-turbo-modules.md

```diff
@@ -0,0 +1,185 @@
+
+The `TurboModuleWrapper` object is a base class that is supposed to be extended by a companion object for TurboModules. The base class has this interface:
+```objc
+@interface TurboModuleWrapper: NSObject
+- (std::shared_ptr<facebook::react::TurboModule>)getTurboModule:
+    (const facebook::react::ObjCTurboModule::InitParams &)params;
+@end
+```
+And the base class implementation just fails, as we don't want for it to be used directly.
+
+Then, the `RCTTurboModule` interface can ask each TM to actually return an implementation of the `TurboModuleWrapper` rather then the actual `TurboModule`.
+
+The `TurboModuleWrapper` is a pure Objective-C class, so it will work seamlessly with ObjectiveC and Swift.
+
+So, now, the public interface of a TurboModule will only have pure objc entries and no C++ code.
+
+When it comes to the implementation, the user-defined TurboModule  won't have to deal with any C++ code:
+
+```objc
+@implementation MyTurboModule
+- (TurboModuleWrapper *)getWrapper
+{
+return [[MyTurboModuleWrapper alloc] init];
+}
+
+// ... rest of the TM methods ...
+
+@end
+```
+
+The `MyTurboModuleWrapper` implementation can be Codegenerated! 
+
+We don't even need to ask our users to write that code themselves, as we have all the informations we need in the Codegen already.
+
+
+The implementation will look like (note that `<UserDefinedName>` is something we get from Codegen):
+```objc
+// In the .h file
+
+@interface Native<UserDefinedName>Wrapper: TurboModuleWrapper
+@end
+
+// In the .mm file
+
+@implementation Native<UserDefinedName>Wrapper
+
+- (std::shared_ptr<facebook::react::TurboModule>)getTurboModule:
+   (const facebook::react::ObjCTurboModule::InitParams &)params;
+{
+  return std::make_shared<facebook::react::Native<UserDefinedName>SpecJSI>(params);
+}
+
+@end
+```
+
+
+The user-defined TurboModule has already access to this file, and so the switch from `getTurboModule` to `getWrapper` doesn't require any additional includes.
+
+
+Finally, we will have to update the `RCTTurboModuleManager` to take this new object into consideration. So we have to modify the provide Turbomodule with the following code
+```objc
+  // Step 2e: Return an exact sub-class of ObjC TurboModule
+  std::shared_ptr<facebook::react::TurboModule> turboModule = nullptr;
+  
+  if ([module respondsToSelector:@selector(getTurboModule:)]) {
+    turboModule = [module getTurboModule:params];
+  } else if ([module respondsToSelector:@selector(getWrapper)]) {
+    auto wrapper = [module getWrapper];
+    turboModule = [wrapper getTurboModule:params];
+  }
+```
+
+
+Additionally we need to make sure that `React_Codegen` module is compatible with importing to Swift. I did a small test and adding few ifdefs to React_Codegen headers allows us to use Swift.
+
+```swift
+import protocol React_Codegen.NativeSwiftTestLibrarySpec
+import protocol React_Codegen.TurboModuleWrapper
+import class React_Codegen.NativeSwiftTestLibraryWrapper
```

Comment:
```suggestion
import protocol ReactCodegen.NativeSwiftTestLibrarySpec
import protocol ReactCodegen.TurboModuleWrapper
import class ReactCodegen.NativeSwiftTestLibraryWrapper
```

Why not just importing `ReactCodegen`?

the `import protocol` and `import class` are kind of an advanced usage of the import statement in swift. Not many people use it or know what they are actually doing.

It could be that importing the whole `ReactCodegen` is not possible as it will make the build fail.

Reply:
Exactly, at this point importing whole module is not possible but that's something we can adjust.